src/services/probes.py

```diff
@@ -0,0 +1,257 @@
+                logger.error(f"{name} connection has an error: {e}")
+                all_ready = False
+
+        return all_ready
+
+    def get_llms_states(self) -> dict[str, bool]:
+        """
+        Get the readiness states of all LLMs.
+
+        Returns:
+            dict[str, bool]: A dictionary where keys are LLM names and values
+            are their readiness states.
+        """
+        self.are_llms_ready()
+        return self._model_states or {}
+
+
+def get_llm_readiness_probe() -> Generator[LLMReadinessProbe, None, None]:
+    """
+    Create and yield an LLMReadinessProbe instance.
+
+    This function initializes the required models using the ModelFactory
+    and yields an LLMReadinessProbe instance. It ensures proper error
+    handling in case of initialization failure.
+    """
+    try:
+        probe = LLMReadinessProbe()
+        if not probe.has_models():
+            probe.set_models(ModelFactory(get_config()).create_models())
+        yield probe
+    except Exception as e:
+        logger.exception(f"Failed to initialize LLMReadinessProbe: {e}")
+        raise
+
+
+def is_hana_ready(connection: IHanaConnection | None) -> bool:
+    """
+    Check if the HANA database is ready.
+
+    Returns:
+        bool: True if HANA is ready, False otherwise.
+    """
+    if not connection:
+        logger.warning("HANA DB connection is not initialized.")
+        return False
+
+    try:
+        if connection.isconnected():
+            logger.info("HANA DB connection is ready.")
+            return True
+    except Exception as e:
+        logger.error(f"Error while connecting to HANA DB: {e}")
+        return False
+    logger.info("HANA DB connection is not ready.")
+    return False
+
+
+def is_redis_ready(connection: IRedisConnection | None) -> bool:
+    """
+    Check if the Redis service is ready.
+
+    Returns:
+        bool: True if Redis is ready, False otherwise.
+    """
+    if not connection:
+        logger.error("Redis connection is not initialized.")
+        return False
+
+    try:
+        if connection.ping():
+            logger.info("Redis connection is ready.")
+            return True
+    except Exception as e:
+        logger.error(f"Redis connection failed: {e}")
+        return False
+    logger.info("Redis connection is not ready.")
+    return False
+
+
+def is_usage_tracker_ready(custom_metrics: CustomMetrics | None) -> bool:
+    """
+    Check if the collection of usage metrics is working without failures.
+
+    Returns:
+        bool: True if Redis is ready, False otherwise.
+    """
+    # If there is no custom metrics, we cannot detect problems so we are
+    # already in a failed state:
+    if not custom_metrics:
+        return False
+
+    # If the metric is None, we are fine.
+    failure_metric_value = custom_metrics.registry.get_sample_value(FAILURE_METRIC_KEY)
+    if not failure_metric_value:
+        return True
+
+    return failure_metric_value == 0.0
```

Comment:
What will happen if the failure happens and recovers? Will it still return `false` because the key exists? IMO, if you want to use this logic then we should be checking if the failure rate is increasing or not? If the failure rate is increasing then it means something is broken.

Reply:
Good point but how would we measure if the usage tracking is recovering? Because this already will measure if the metric increases. Do you have some kind of threshold or threshold per time in mind? I find it hard to predict the behaviour of the metrics and to abstract some kind of recovery logic out of it.